PhotoGalleryXamarin/Fragments/PhotoGalleryFragment.cs

```diff
@@ -56,35 +62,54 @@ public void SetupAdapter()
             }
         }
 
-        private void OnItemsFetched(List<GalleryItem> galleryItems)
+        private void GlobalLayout() 
         {
-            _galleryItems = galleryItems;
-            SetupAdapter();
-        }
+            _columnNumber = _recyclerView.Width / ColWidth;
+            _recyclerView.SetLayoutManager(new GridLayoutManager(Activity, _columnNumber));
+        }
 
-        private class PhotoAdapter : RecyclerView.Adapter
+        private void ScrollChanged(object sender, View.ScrollChangeEventArgs e)
         {
-            private List<GalleryItem> _galleryItems;
+            if (!_recyclerView.CanScrollVertically(1))
+            {
+                _pageIndex++;
+                new FetchItemsTask
+                {
+                    ResultPageIndex = _pageIndex,
+                    OnPostExecuteImpl = OnItemsFetched
+                }.Execute();
+            }
+        }
 
-            public override int ItemCount => _galleryItems.Count;
+        private void OnItemsFetched(List<GalleryItem> galleryItems)
+        {
+            var previousSize = _galleryItems.Count;

```

Comment:
let's rename this variable to indexToInsert, as it would better comunicate its intent